src/lib.rs

```diff
@@ -143,53 +100,178 @@ impl Default for Options {
+// we provide it by wrapping the writer given to us while escaping appropriately any string to
+// be written, depending on the type of node we're writing.
+struct FmtWriter<W: Write> {
+    writer: W,
+    error_kind: Option<io::ErrorKind>,
+    escape: Option<Escape>,
+}
+
+impl<W: Write> FmtWriter<W> {
+    fn take_err(&mut self) -> io::Error {
+        let error_kind = self
+            .error_kind
+            .expect("there must have been an error before calling take_err()!");
+        // This avoids forgetting to set it to the appropriate value when calling write_fmt().
+        // We can't do it in FmtWriter's write_str(), since with a real format string the method
+        // will be called several times so it'll fail in the expect() below as we'll have set
+        // self.escape back to None.
+        self.escape = None;
+        // Make sure we can detect if take_err() is called without having an error that happened beforehand
+        self.error_kind = None;
+
+        // There's just no way of properly copying the io::Error (no Copy or Clone available), so
+        // we have no choice to create a new one, which likely loses the backtrace up to this point.
+        io::Error::from(error_kind)
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+enum Escape {
+    Comment,
+    AttributeValue,
+    Text,
+}
+
+fn write_escaped<W: Write>(
+    writer: &mut W,
+    s: &str,
+    // Don't bother passing around XmlWriter.opt.use_single_quote just for that: just
+    // always escape both double and single quotes, whatever quoting character we might
+    // be using for attributes.
+    escape_quotes: bool,
+) -> io::Result<()> {
+    let mut part_start_pos = 0;
+    for (byte_pos, byte) in s.bytes().enumerate() {
+        let escaped_char: Option<&[u8]> = match byte {
+            b'&' => Some(b"&amp;"),
+            b'>' => Some(b"&gt;"),
+            b'<' => Some(b"&lt;"),
+            b'"' if escape_quotes => Some(b"&quot;"),
+            b'\'' if escape_quotes => Some(b"&apos;"),
+            _ => None,
+        };
+        if let Some(escaped_char) = escaped_char {
+            // We have a character to escape, so write the previous part and the escaped character
+            writer.write_all(&s[part_start_pos..byte_pos].as_bytes())?;
+            writer.write_all(escaped_char)?;
+            // +1 skips the escaped character from part, for afterwards
+            part_start_pos = byte_pos + 1;
+        }
+        // There's nothing to be done if the character doesn't need to be escaped, as we'll either
+        // wait until we get an escapable character, or wait until the end of the string where we'll
+        // just write out the rest of the string.
+    }
+    // Write the rest of the string which needs no escaping
+    writer.write_all(&s[part_start_pos..].as_bytes())
+}
+
+fn write_escaped_comment<W: Write>(writer: &mut W, s: &str) -> io::Result<()> {
```

Comment:
Shouldn't we simply return an error or silently skip `--` to begin with?

Reply:
I think I will just not escape comments at all. The industry standard for XML parsing/writing, libxml2, doesn't even do it, and neither do libxml2/roxmltree/quick-xml accept comments with `--`, so it's not possible to reach the point where you'd be writing invalid comments when writing back some parsed XML. As such the only time where you'd write comments would be when hard-coding some in your code, as I doubt anyone write user-inputed data to comments (that's just not how XML is to be used…).

In any case, I believe escaping comments is not needed so I'll just drop this escaping